### ISSUE TITLE
Unwinding for Branches and Loops

### ISSUE THREAD

**Issue by ahorn** [losfair/wasm-core#15]
Could I ask you a quick question about the [Executor](https://github.com/losfair/wasm-core/blob/23fdad12cfceb3a91b9f7d615678b7ac0cb144b4/src/executor.rs)?

It handles (un)conditional jumps by adjusting the instruction pointer (`ip`). The specification for [control instructions](https://webassembly.github.io/spec/core/syntax/instructions.html#control-instructions) says that "[t]aking a branch *unwinds* the operand stack up to the height where the targeted structured control instruction was entered". This unwinding is [formalized](https://webassembly.github.io/spec/core/exec/instructions.html#control-instructions) by repeatedly popping the operand stack according to the provided label index.

Where is this unwinding happening in wasm-core? Thank you for your time!

**Comment by losfair** [comment]
It's definitely incorrect :(

Actually I was aware of this while writing the [SSA analyzer](https://github.com/losfair/wasm-core/blob/master/src/ssa.rs). Any code affected by the bug should be rejected by the analysis due to [lcss](https://github.com/losfair/wasm-core/blob/master/src/ssa.rs#L380) unable to figure out the new stack state, though I haven't seen any errors of this kind from llvm-generated wasm binaries.

I think this can be fixed by adding explicit stack state information in `Jmp*` instructions in the translation phase and making SSA analysis use it instead of figuring out everything itself. (and also, deprecate the heavily flawed interpreter)

**Comment by ahorn** [comment]
Thanks for sharing your thoughts.

When you describe the interpreter as "heavily flawed", are you thinking of the unwinding bug in particular, or other aspects of the design, too? Are you implying the interpreter is to be replaced entirely?

**Comment by losfair** [comment]
Mainly for performance reasons:

- heap allocation everywhere
- runtime type checking
- ...

The current interpreter absolutely needs a rewrite, and I'm thinking about different approaches:

- stack machine like the current one, but performs type erasure and unwinds the stack correctly :)
- generate & execute register-machine-style bytecode from the SSA graph

What's your thoughts on this?

**Comment by ahorn** [comment]
Yes, after [validation](https://webassembly.github.io/spec/core/valid/index.html), ignoring types should be okay, except for the general case of `indirect_call` etc. 

For a possible re-design, I would want to review existing Wasm interpreters. For example, the [V8 Wasm interpreter](https://github.com/v8/v8/blob/785bd43b7e15fba044b2ed1323bb0018e2c193fa/src/wasm/wasm-interpreter.cc) implements a stack machine. Do you think such an implementation is inherently more inefficient than a register-machine style interpreter?
